integration/commands_administration_test.go

```diff
@@ -773,6 +775,227 @@ func TestCommandsAdministrationDBStatsEmptyWithScale(t *testing.T) {
 }
 
+func TestCommandsAdministrationRenameCollection(t *testing.T) {
+	t.Parallel()
+
+	maxTableLen := strings.Repeat("a", 63) // maxTableLen = 63
+	insertCollections := []string{"foo", "buz"}
+
+	for name, tc := range map[string]struct { //nolint:vet // for readability
+		collection       string
+		targetCollection string
+		expected         bson.D
+		err              *mongo.CommandError
+		toInsert         []interface{}
+	}{
+		"Rename": {
+			collection:       "foo",
+			targetCollection: "bar",
+			expected:         bson.D{{"ok", float64(1)}},
+		},
+		"RenameSame": {
+			collection:       "foo",
+			targetCollection: "foo",
+			err: &mongo.CommandError{
+				Code:    20,
+				Name:    "IllegalOperation",
+				Message: `Can't rename a collection to itself`,
+			},
+		},
+		"RenameDuplicate": {
+			collection:       "foo",
+			targetCollection: "buz",
+			err: &mongo.CommandError{
+				Code:    48,
+				Name:    "NamespaceExists",
+				Message: `target namespace exists`,
+			},
+		},
+		"SoureDoesNotExist": {
+			collection:       "none",
+			targetCollection: "bar",
+			err: &mongo.CommandError{
+				Code:    26,
+				Name:    "NamespaceNotFound",
+				Message: "source namespace does not exist",
+			},
+		},
+		// this confirms that after we rename foo to bar and then create foo again,
+		// 1. bool-false doesn't exist
+		// 2. the newly inserted documents exist
+		"InsertIntoOld": {
+			collection:       "foo",
+			targetCollection: "bar",
+			expected:         bson.D{{"ok", float64(1)}},
+			toInsert: []interface{}{
+				bson.D{{"_id", 1}},
+				bson.D{{"_id", 2}},
+			},
+		},
+		"RenameMaxTableLen": {
+			collection:       "foo",
+			targetCollection: maxTableLen + "a",
+			err: &mongo.CommandError{
+				Code:    73,
+				Name:    "InvalidNamespace",
+				Message: "collection is too long",
+			},
+		},
+	} {
+		name, tc := name, tc
+		t.Run(name, func(t *testing.T) {
+			t.Parallel()
+
+			ctx, collection := setup.Setup(t, shareddata.Bools)
+
+			db := collection.Database()
+
+			require.NotEmpty(t, insertCollections)
+			for _, coll := range insertCollections {
+				require.NoError(t, db.CreateCollection(ctx, coll))
+			}
+
+			var actual bson.D
+
+			sourceNamespace := fmt.Sprintf("%s.%s", db.Name(), tc.collection)
+			targetNamespace := fmt.Sprintf("%s.%s", db.Name(), tc.targetCollection)
+
+			cmd := bson.D{
+				{"renameCollection", sourceNamespace},
+				{"to", targetNamespace},
+			}
+
+			err := collection.Database().Client().Database(db.Name()).RunCommand(
+				ctx, cmd,
+			).Decode(&actual)
+
+			if tc.err != nil {
+				AssertEqualError(t, *tc.err, err)
+				return
+			}
+
+			require.NoError(t, err)
+			require.Equal(t, tc.expected, actual)
+
+			collections, err := db.ListCollectionNames(
+				ctx,
+				bson.D{},
+				nil,
+			)
+
+			require.NoError(t, err)
+
+			require.Contains(t, collections, tc.targetCollection)
+			require.NotContains(t, collections, tc.collection)
+
+			if tc.toInsert != nil {
```

Comment:
I think this is quite confusing, parameter is `tc.toInsert` but what is it doing is creating a new collection.
My suggestion is having `tc.newCollection` or other better name and  `if tc.newCollection != ""` condition to insert.

Also make sure to document that newCollection is a collection created after renaming collection.

Reply:
I see what you are saying but we re-created the old collection so I chose a different approach, to add a new field `recreateOld` which triggers the insertion and verifies that the metadata is new.